### ISSUE TITLE
Smart rounding

### ISSUE THREAD

**Issue by jduerholt** [experimental-design/bofire#146]
@DavidWalz  cc @jkleinekorte and @bertiqwerty 

I thought about rounding functionalities in bofire for example in the context of using continuous relaxations for NChooseK as described in https://github.com/experimental-design/bofire/issues/145 which could lead to subtle numeric violations of the original constraint. Ideally we would have a rounding function which takes as input the domain with all features and constraints in addition to the generated candidate and then rounds it to the closest feasible solution.

Another relevant example would be if would introduce something like precision or step size for the `ContinuousInput`, especially if it used in combination with some kind of linear constraint and a direct rounding to the next allowed level would lead to constraint violations. 

Could one not describe this rounding problem as a mixed integer optimization problem where we want to minimize the (squared) distance between a rounded feasible solution and the actual proposed candidate under the provided constraints and discretization levels?

What do you think?

**Comment by DavidWalz** [comment]
Yes, restoring feasibility for {NChooseK + linear constraints} with L2 distance to the orginal point as objective is an MIQP. 
We can pull in an additional dependecy to do this. `python-mip` and `cvxpy` (with CBC as solver) are lightweight, but only support LPs, so the L2^2 function would need to be linearized. `pyomo` also allows for formulating QPs and apparently CBC can solve them, but `pyomo` might be a bit heavy.  Alternatively an L1 distance could be used which only involves max.
Restoring feasibility for nonlinear constraints wouldn't work.
Is that what you have in mind?

There are some space filling designs that can be formulated as MIPs and certain surrogates could be optimized globally (like is done in ENTMOOT), so in general I like the idea of extending bofire in that direction.

**Comment by DavidWalz** [comment]
ping @R-M-Lee

**Comment by jduerholt** [comment]
Yes, I have a MIQP in mind and more specifically a function which gets as input a domain, and a candidate and returns the rounded solution ;) I am not an expert in this kind of optimization, but could we not use `cvxpy` in combination with `pyscipopt`, as described [here](https://www.cvxpy.org/examples/basic/mixed_integer_quadratic_program.html)?

We should then also add a preciscion attribute to the ContinuousInputFeauture.

**Comment by DavidWalz** [comment]
Nice. A drawback is that `pyscipopt` does not provide wheels so it's again conda (at least for windows) or getting scip compiled in our containerized backends. Given that it's the same for `ipopt` that's ok. It just means that the smart rounding would be optional and can't be relied on to be present.

**Comment by jduerholt** [comment]
This would be fine for me ...

**Comment by rosonaeldred** [comment]
https://github.com/experimental-design/bofire/issues/166
